### ISSUE TITLE
Robust HTTP Server Example

### ISSUE THREAD

**Issue by vinniefalco** [boostorg/beast#2205]
It would be nice to have a new example HTTP server that supports POST to upload files, GET / to see the list of files, and GET <file> to download a file.

**Comment by Tectu** [comment]
I am currently working on that...
Not that anybody should necessarily wait for that. My experience working with beast is very limited at the moment.

**Comment by Tectu** [comment]
The current code (on my end) is based on C++20 (could be easily down-graded to C++17). Does this even make for a useful example for you guys?

**Comment by madmongo1** [comment]
Personally I think there is great value in a c++20 example. 
@vinniefalco I presume we'll also want to implement a c++11 version?

**Comment by vinniefalco** [comment]
You have to use your best judgement. Try to follow the existing examples in terms of style. A C++20 example is okay (I assume because you want to use `co_await`). But in that case you also need a regular async/callback port of it. We can leave out the synchronous version :)

**Comment by Tectu** [comment]
Here is my "example" on how to use `boost.beast`: https://github.com/Tectu/malloy
I might have digressed a bit.
Maybe it's still worth listing it somewhere?

This is so far mainly the grunt work. It has many rough edges and needs serious improvements in several places.

I'd be thankful for any kind of feedback (or direct contributions!)

**Comment by Tectu** [comment]
The "example" has been released as `v0.1.0`: https://github.com/Tectu/malloy

Current feature list:

- High-level controller to setup I/O context, SSL context, worker threads and more
- HTTP
  - Plain or TLS (SSL) connections
  - Cookies
  - Sessions
  - Upgrading connections to WebSocket
  - Client
    - Response filters
  - Server
    - Routing
      - Simple handlers (useful for building REST APIs)
        - Target matching via regex
        - Capturing groups via regex
      - Sub-routers (nested/chained routers)
      - Redirections
      - File serving locations
      - Preflight responses
      - Access policies
        - HTTP basic auth
        - Custom access policies
      - Websocket endpoints (with auto-upgrade from HTTP)
    - Request filters
- WebSocket
  - Client
  - Server
  - Connections upgradable from HTTP
- HTML
  - Forms
    - Supported encoding types
      - `application/x-www-form-urlencoded`
      - `multipart/form-data`
      - `text/plain`
    - Parsing
    - Rendering

Feedback and contributions are highly welcomed.

**Comment by vinniefalco** [comment]
This is BIG !!

**Comment by vinniefalco** [comment]
This is more like a framework than an example

**Comment by Tectu** [comment]
> This is more like a framework than an example

Hence I wrote _"example"_ in quotation marks :p

No but seriously. This honestly started off as an example - I kid you not. I had several application using `beast` and a lot of the stuff was being repeated between applications. Putting this into a library/framework made sense (to me).

I will keep push this forward. I think that a fair share of `beast` users are dealing with lower level stuff they don't necessarily want to deal with but have to. I hope that this library will ease the pain and provide developers looking for a higher-level solution with a solid base built on top of `beast`.

**Comment by vinniefalco** [comment]
I was so caught up looking at the wonderful work you've gone that I forgot to Star, Like and smash that Subscribe button

**Comment by Tectu** [comment]
Haha ^^

So, do you think this is useful to you and other `beast` users? Should we point people towards this as an option for those who'd like to avoid the lower level stuff?

**Comment by madmongo1** [comment]
I’d like to write a little application with it to give you some feedback.

**Comment by Tectu** [comment]
Please do - that would be highly appreciated.

**Comment by Tectu** [comment]
New release is available: https://github.com/Tectu/malloy/releases/tag/0.3.0

**Comment by Tectu** [comment]
Seems like we're on the right track here, guys!
Some feedback from a user & contributor of malloy:

> Serving a file with load balanced across multiple threads and asynchronous in less than 50 lines is damn impressive (its one of the reasons I started using and contributing to malloy, the router and client interfaces are really nice well done :p)

I think this approach of `boost.beast` being low level and `malloy` being higher level is gonna work out in everyone's favor!
As such: Please keep up the good work guys!

**Comment by madmongo1** [comment]
Wonderful.

**Comment by klemens-morgenstern** [comment]
@Tectu any progress to report?

**Comment by Tectu** [comment]
> @Tectu any progress to report?

Yeah, lots of progress to report - anything specific you're interested in?
The next release of malloy (`0.6.0`) is expected next week.
There is also work ongoing for providing some higher-level interface to create REST APIs in the `feature/rest` branch. By now, malloy has also been used quite successfully in a couple of real-world scenarios.

**Comment by vinniefalco** [comment]
Beast2 is in development:

https://github.com/cppalliance/beast2
https://github.com/cppalliance/http
https://github.com/cppalliance/corosio
https://github.com/cppalliance/capy
